include/ensmallen_bits/bigbatch_sgd/adaptive_stepsize.hpp

```diff
@@ -87,6 +87,55 @@ class AdaptiveStepsize
     // Update the iterate.
     iterate -= stepSize * gradient;
 
+    // Update Gradient & calculate curvature of quadratic approximation.
+    arma::mat functionGradient(iterate.n_rows, iterate.n_cols);
+    arma::mat gradPrevIterate(iterate.n_rows, iterate.n_cols);
+    arma::mat functionGradientPrev(iterate.n_rows, iterate.n_cols);
+
+    double vB = 0;
+    arma::mat delta0, delta1;
+
+    // Initialize previous iterate, if not already initialized.
+    if (iteratePrev.is_empty())
+    {
+      iteratePrev.zeros(iterate.n_rows, iterate.n_cols);
+    }
+
+    // Compute the stochastic gradient estimation.
+    function.Gradient(iterate, offset, gradient, 1);
+    function.Gradient(iteratePrev, offset, gradPrevIterate, 1);
+
+    delta1 = gradient;
+
+    for (size_t j = 1, k = 1; j < backtrackingBatchSize; ++j, ++k)
+    {
+      function.Gradient(iterate, offset + j, functionGradient, 1);
+      delta0 = delta1 + (functionGradient - delta1) / k;
+
+      // Compute sample variance.
+      vB += arma::norm(functionGradient - delta1, 2.0) *
+          arma::norm(functionGradient - delta0, 2.0);
+
+      delta1 = delta0;
+      gradient += functionGradient;
+
+      // Used for curvature calculation.
+      function.Gradient(iteratePrev, offset + j, functionGradientPrev, 1);
```

Comment:
Looks like we could move the gradient outside the for loop and use ``backtrackingBatchSize`` instead of ``1``.

Reply:
Right, we can do that.